File: rust/src/leetcode/editor/cn/flatten_binary_tree_to_linked_list.rs
use crate::leetcode::editor::cn::util::*;
struct Solution;
//leetcode submit region begin(Prohibit modification and deletion)
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::ops::Deref;

impl Solution {
    pub fn flatten(root: &mut Option<Rc<RefCell<TreeNode>>>) {
        match root {
            None => {}
            Some(root) => {
                let mut r = root.borrow_mut();
                Self::flatten(&mut r.left);
                Self::flatten(&mut r.right);

                let left = r.left.take();
                let right = r.right.take();
                r.right = left;

                let mut cur = root.clone();
                while let Some(ref cur_ref) = cur.clone().borrow().right {
                    cur = cur_ref.clone();
                }
                cur.borrow_mut().right = right;
            }
        }
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: jvm/leetcode_java/src/main/java/leetcodecn/FlattenBinaryTreeToLinkedList.java
package leetcodecn;

// [114]二叉树展开为链表
class FlattenBinaryTreeToLinkedList {


//IMPORTANT!! Submit Code Region Begin(Do not remove this line)

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    class Solution {
        public void flatten(TreeNode root) {
            if (root != null) {
                traverse(root);
            }
        }

        private TreeNode traverse(TreeNode node) {
            TreeNode leftEnd = null;
            TreeNode rightEnd = null;
            TreeNode left = node.left;
            TreeNode right = node.right;

            if (left != null) {
                leftEnd = traverse(left);
            }
            if (right != null) {
                rightEnd = traverse(right);
            }

            // 左子树不为空，那么将左子树的尾部连接到右子树
            // 否则因为右子树本来就连接着，所以啥都不用做
            if (left != null) {
                node.right = left;
                leftEnd.right = right;
            }
            node.left = null;

            return (rightEnd != null) ? rightEnd : ((leftEnd != null) ? leftEnd : node);
        }
    }
//IMPORTANT!! Submit Code Region End(Do not remove this line)

    public static void main(String[] args) {
        // add your test code
    }
}
